api/src/main/java/me/lokka30/treasury/api/economy/account/Account.java

```diff
@@ -77,46 +79,109 @@ void setBalance(
      *
      * <p>Specified amounts must be ABOVE zero.
      *
-     * @param amount        the amount the balance will be reduced by
-     * @param initiatorUUID the one who initiated the transaction
-     * @param currency      the {@link Currency} of the balance being modified
-     * @param subscription  the {@link EconomySubscriber} accepting the new balance
+     * @param amount       the amount the balance will be reduced by
+     * @param initiator    the one who initiated the transaction
+     * @param currency     the {@link Currency} of the balance being modified
+     * @param subscription the {@link EconomySubscriber} accepting the new balance
      * @author lokka30, Geolykt, MrIvanPlays
-     * @see Account#setBalance(double, UUID, Currency, EconomySubscriber)
+     * @see Account#setBalance(double, EconomyTransactionInitiator, Currency, EconomySubscriber)
      * @see Account#doTransaction(EconomyTransaction, EconomySubscriber)
      * @since {@link me.lokka30.treasury.api.economy.misc.EconomyAPIVersion#v1_0 v1.0}
      */
     default void withdrawBalance(
             double amount,
-            @NotNull UUID initiatorUUID,
+            @NotNull EconomyTransactionInitiator<?> initiator,
             @NotNull Currency currency,
             @NotNull EconomySubscriber<Double> subscription
     ) {
-        doTransaction(EconomyTransaction.newBuilder().withCurrency(currency).withInitiator(initiatorUUID).withTransactionAmount(
-                amount).withTransactionType(EconomyTransactionType.WITHDRAWAL).build(), subscription);
+        withdrawBalance(amount, initiator, currency, null, subscription);
+    }
+
+    /**
+     * Withdraw an amount from the {@code Account} balance.
+     *
+     * <p>Specified amounts must be ABOVE zero.
+     *
+     * @param amount       the amount the balance will be reduced by
+     * @param initiator    the one who initiated the transaction
+     * @param currency     the {@link Currency} of the balance being modified
+     * @param reason       the reason of why the balance is modified
+     * @param subscription the {@link EconomySubscriber} accepting the new balance
+     * @author MrIvanPlays
+     * @see Account#setBalance(double, EconomyTransactionInitiator, Currency, EconomySubscriber)
+     * @see Account#doTransaction(EconomyTransaction, EconomySubscriber)
+     * @since {@link me.lokka30.treasury.api.economy.misc.EconomyAPIVersion#v1_0 v1.0}
+     */
+    default void withdrawBalance(
+            double amount,
+            @NotNull EconomyTransactionInitiator<?> initiator,
+            @NotNull Currency currency,
```

Comment:
I have commented this several times but, wouldn't it be better to have as parameter the UUID of the currency instead of the instance itself?

Reply:
```java
withdrawBalance(20, initiator, currency, // ...
// vs
withdrawBalance(20, initiator, currency.getCurrencyId(), // ...
```
BUT WAIT THERE'S MORE TO THAT!!!! Every time the providing plugin needs a Currency instance it would have to query it from the CurrencyManager, whilst for the depending plugin it is easier, because it can store the `Currency` object somewhere static

To answer your question, no.

Reply:
I forgot that CurrencyManager returns an `Optional<Currency>`, I thought we would have to use a subscriber to get a currency so it would get messy fast. 
Now that I've checked the currency manager, there is really no diference between them so I don't mind either option

Reply:
Is this conversation resolved?

Reply:
Yes.